### ISSUE TITLE
Should `Complex` be convertible to `Quaternion`?

### ISSUE THREAD

**Issue by hyrodium** [JuliaGeometry/Quaternions.jl#62]
We already have constructor and `convert` method from `Complex` to `Quaternion`.
```julia
julia> Quaternion(Complex(1,2))
Quaternion{Int64}(1, 2, 0, 0, false)

julia> convert(Quaternion, Complex(1,2))
Quaternion{Int64}(1, 2, 0, 0, false)
```
However, we still missing the following methods `+` and `*`.
```julia
julia> Quaternion(1,2,3,4) + Complex(1,2)
ERROR: MethodError: no method matching imag(::Complex{Int64})
You may have intended to import Base.imag
Closest candidates are:
  imag(::Quaternion) at ~/.julia/dev/Quaternions/src/Quaternion.jl:44
  imag(::Octonion) at ~/.julia/dev/Quaternions/src/Octonion.jl:48
Stacktrace:
 [1] convert(#unused#::Type{Quaternion{Int64}}, z::Complex{Int64})
   @ Quaternions ~/.julia/dev/Quaternions/src/Quaternion.jl:21
 [2] _promote
   @ ./promotion.jl:327 [inlined]
 [3] promote
   @ ./promotion.jl:350 [inlined]
 [4] +(x::Quaternion{Int64}, y::Complex{Int64})
   @ Base ./promotion.jl:379
 [5] top-level scope
   @ REPL[23]:1

julia> Quaternion(1,2,3,4) * Complex(1,2)
ERROR: MethodError: no method matching imag(::Complex{Int64})
You may have intended to import Base.imag
Closest candidates are:
  imag(::Quaternion) at ~/.julia/dev/Quaternions/src/Quaternion.jl:44
  imag(::Octonion) at ~/.julia/dev/Quaternions/src/Octonion.jl:48
Stacktrace:
 [1] convert(#unused#::Type{Quaternion{Int64}}, z::Complex{Int64})
   @ Quaternions ~/.julia/dev/Quaternions/src/Quaternion.jl:21
 [2] _promote
   @ ./promotion.jl:327 [inlined]
 [3] promote
   @ ./promotion.jl:350 [inlined]
 [4] *(x::Quaternion{Int64}, y::Complex{Int64})
   @ Base ./promotion.jl:380
 [5] top-level scope
   @ REPL[24]:1
```

I think we have the following choices to solve this problem.

* Remove the constructor and `convert` method
    * There is no natural embedding ℂ ⊆ ℍ because it depends on the choice of the basis.
    * This will make a breaking change.
* Add type promotions
    * This will not make a breaking change.

I don't know the practical benefit of converting from `Complex` to `Quaternion`, so I prefer the first.

Related comment: https://github.com/JuliaGeometry/Quaternions.jl/issues/29#issuecomment-732327293

**Comment by sethaxen** [comment]
I agree there is not natural embedding of `ℂ` in `ℍ`, and it would be better if we did not have this conversions. Removing these would be breaking.

But this is orthogonal to the error at the beginning, which is because Quaternions has its own `imag` function that is different from `Base.imag`, and it's calling the wrong with in the `convert` method:
https://github.com/JuliaGeometry/Quaternions.jl/blob/23a8212b1761c200030aa6614c4e0dbf971a0d9f/src/Quaternion.jl#L21-L22
`Base.imag` needs to be explicitly called here. This is repaired in #56, but a separate PR to fix the bug wouldn't be a bad thing.

**Comment by dkarrasch** [comment]
Yes, there has been a quick fix for the conversion issue around in #40 (ignoring the general embedding issue).

**Comment by sethaxen** [comment]
The bug was fixed in #52
